File: content/development/golang-first-impression/worker.cpp
#include <cstdio>
#include <future>
#include <vector>
#include <unistd.h>

int called_from_async(int a) {
  sleep(1);
  printf("async call : %d\n", a);
  return a * 10;
}

int main()
{
  const int loop_count = 3;
  std::vector<std::future<int>> results;

  for(int i = 0 ; i < loop_count ; ++i) {
    std::future<int> result(std::async(called_from_async, i));
    results.push_back(std::move(result));
  }

  printf("generating async complete\n");

  for(auto &result : results) {
    printf("retval : %d\n", result.get());
  }
  return 0;
}

File: content/posts/development/golang-first-impression/maya/worker.go
package main

import (
	"fmt"
	"time"
)

func calledFromAsync(a int, ch chan int) {
	time.Sleep(1 * time.Second)
	fmt.Println("async call : ", a)
	ch <- a * 10
}

func main() {
	loopCount := 3
	ch := make(chan int)

	for i := 0; i < loopCount; i++ {
		go calledFromAsync(i, ch)
	}

	fmt.Println("generating async complete")

	for i := 0; i < loopCount; i++ {
		fmt.Println("retval : ", <-ch)
	}
}
